lib/chef/provider/package/msu.rb

```diff
@@ -0,0 +1,148 @@
+        include Chef::Mixin::ShellOut
+        include Chef::Mixin::Uris
+        include Chef::Mixin::Checksum
+
+        provides :msu_package, os: "windows"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::MsuPackage.new(new_resource.name)
+
+          # download file if source is a url
+          download_source_file if uri_scheme?(new_resource.source)
+
+          # temp directory where the contents of msu file get extracted
+          @temp_directory = Dir.mktmpdir("chef")
+          extract_msu_contents(@new_resource.source, @temp_directory)
+          @cab_files = read_cab_files_from_xml(@temp_directory)
+
+          unless @cab_files.empty?
+            # set the package version
+            # Here is a catch: How to set package version in case of multiple cab files?
+            @new_resource.source = @cab_files.first
+            cab_pkg = Chef::Provider::Package::Cab.new(@new_resource, nil)
+            new_resource.version(cab_pkg.package_version)
+            current_resource.version(cab_pkg.installed_version)
+          else
+            raise Chef::Exceptions::Package, "Corrupt MSU package: MSU package XML doesn't contain any cab file"
+          end
+          current_resource
+        end
+
+        def download_source_file
+          source_resource.run_action(:create)
+          Chef::Log.debug("#{new_resource} fetched source file to #{source_resource.path}")
+          @new_resource.source = source_resource.path
+        end
+
+        def source_resource
+          @source_resource ||= Chef::Resource::RemoteFile.new(default_download_cache_path, run_context).tap do |r|
+            r.source(new_resource.source)
+            r.cookbook_name = new_resource.cookbook_name
+            r.checksum(new_resource.checksum)
+            r.backup(false)
+
+            if new_resource.remote_file_attributes
+              new_resource.remote_file_attributes.each do |(k, v)|
+                r.send(k.to_sym, v)
+              end
+            end
+          end
+        end
+
+        def default_download_cache_path
+          uri = ::URI.parse(new_resource.source)
+          filename = ::File.basename(::URI.unescape(uri.path))
+          file_cache_dir = Chef::FileCache.create_cache_path("package/")
+          Chef::Util::PathHelper.cleanpath("#{file_cache_dir}/#{filename}")
+        end
+
+        def install_package(name, version)
+          #use cab_package resource to install the extracted cab packages
+          @cab_files.each do |cab_file|
+            cab_pkg = Chef::Resource::CabPackage.new(@new_resource.name, run_context)
+            cab_pkg.source = cab_file
+            cab_pkg.run_action(:install)
+          end
+        end
+
+        def remove_package(name, version)
+          #use cab_package provider to remove the extracted cab packages
+          @cab_files.each do |cab_file|
+            cab_pkg = Chef::Resource::CabPackage.new(@new_resource.name, run_context)
```

Comment:
same here that these two cab package resources should use `declare_resource(:cab_package, ....)`